### PR TITLE
Make type inference failures more consistent

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/mypy/issues/6998

This PR makes several related changes:
* Reserve `Cannot determine type` error for cases where node deferral failed etc (for example with runtime recursive definitions); don't show it after failed (ambiguous) type inference.
* Always set reasonable variable type after failed inference (like `List[Any]` for an empty list).
* Always give `Need type annotation for variable` error, not just for instances and tuples (that was ad-hoc IMO)

This may actually give some _new_ errors as compared to status quo, but as one can see from the diff, the result is negative in number of (redundant) errors.

There are couple changes in tests that look unrelated, these are because `is_same_type(..., UnboundType())` returns `True` for everything, including `UninhabitedType`. I would say we should actually avoid leaking unbound types from semantic analysis and replace the, with `Any`, but this is a separate issue.

### CODE REVIEW

File: test-data/unit/check-generics.test [hunk 1]
```diff
@@ -996,7 +996,7 @@ IntNode[int](1, 1)
 IntNode[int](1, 'a')  # E: Argument 2 to "Node" has incompatible type "str"; expected "int"
 
 SameNode = Node[T, T]
-ff = SameNode[T](1, 1)  # E: Need type annotation for 'ff'
+ff = SameNode[T](1, 1)
```
Comment (JukkaL) [inline test-data/unit/check-generics.test]: I think that this should generate an error, since `T` is not in scope here. If this seems unrelated, you can create an issue about this (and add a TODO comment here).

File: test-data/unit/check-newsemanal.test [hunk 2]
```diff
@@ -2716,3 +2716,34 @@ class N(NamedTuple):
         )
         b
 [builtins fixtures/tuple.pyi]
+
+[case testNewAnalyzerLessErrorsNeedAnnotation]
```
Comment (JukkaL) [inline test-data/unit/check-newsemanal.test]: What about adding a test case for `x = []  # type: ignore` followed by `reveal_type(x)`?

File: mypy/checker.py [hunk 3]
```diff
@@ -4300,26 +4298,24 @@ def is_valid_inferred_type(typ: Type) -> bool:
...
-            if not is_valid_inferred_type_component(item):
-                return False
-    return True
+    def visit_uninhabited_type(self, t: UninhabitedType) -> bool:
+        return t.ambiguous
+
+
+class SetNothingToAny(TypeTranslator):
+    """Replace all ambiguous <nothing> types with Any (to avoid spurious extra errors)."""
+    def visit_uninhabited_type(self, t: UninhabitedType) -> Type:
```
Comment (JukkaL) [inline mypy/checker.py]: Style nit: add empty line after class docstring.

File: mypy/checker.py [hunk 4]
```diff
@@ -4300,26 +4298,24 @@ def is_valid_inferred_type(typ: Type) -> bool:
...
         # leave_partial_types when we pop a partial types scope.
         return False
-    return is_valid_inferred_type_component(typ)
+    return not typ.accept(NothingSeeker())
 
 
-def is_valid_inferred_type_component(typ: Type) -> bool:
-    """Is this part of a type a valid inferred type?
+class NothingSeeker(TypeQuery[bool]):
+    """Find any <nothing> types resulting from failed (ambiguous) type inference."""
```
Comment (JukkaL) [inline mypy/checker.py]: Style nit: add empty line after class docstring.

File: test-data/unit/fine-grained-cycles.test [hunk 5]
```diff
@@ -205,8 +205,7 @@ def h() -> None:
 [out]
 ==
 a.py:3: error: Invalid type "b.C"
-b.py:6: error: Need type annotation for 'c'
-b.py:7: error: Cannot determine type of 'c'
+b.py:7: error: C? has no attribute "g"
```
Comment (JukkaL) [inline test-data/unit/fine-grained-cycles.test]: Hmm do we now propagate unbound types in type inference? Not sure if that's a good or bad thing.

Comment (ilevkivskyi) [reply test-data/unit/fine-grained-cycles.test]: > Not sure if that's a good or bad thing.

This is rather bad, but I believe the right solution is https://github.com/python/mypy/issues/4987.

**Comment by ilevkivskyi** [comment]
> I still don't fully understand what caused the different behavior in new semantic analyzer. Could you explain that in some detail?

See the discussion in PR https://github.com/python/mypy/pull/6450 and the change in `[case testUnderspecifiedInferenceResult]` there.
